src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,645 @@
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertInForEach(source, out documentUpdate);
+                    case SyntaxKind.EqualsValueClause:
+                    case SyntaxKind.SimpleAssignmentExpression:
+                    case SyntaxKind.MultiplyAssignmentExpression:
+                        return TryReplaceWithLocalFunction(source, out documentUpdate);
+                    case SyntaxKind.ParenthesizedExpression:
+                        return TryConvertInParenthesizedExpression(source, out documentUpdate);
+                    default:
+                        documentUpdate = default;
+                        return false;
+                }
+            }
+
+            private bool TryConvertInParenthesizedExpression(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                // TODO any other cases?
+                var parenthesizedExpression = (ParenthesizedExpressionSyntax)source.Parent;
+                if (parenthesizedExpression.Parent.Kind() == SyntaxKind.SimpleMemberAccessExpression)
+                {
+                    var memberAccessExpression = (MemberAccessExpressionSyntax)parenthesizedExpression.Parent;
+
+                    Func<ExpressionSyntax, ExpressionSyntax, ExpressionSyntax> insideLoopExpression;
+                    ExpressionSyntax initializer;
+                    string variableNamePrefix;
+                    switch (memberAccessExpression.Name.Identifier.ValueText)
+                    {
+                        case nameof(Enumerable.ToList):
+                            if (TryGetTypeSyntax(source, out var typeSyntax))
+                            {
+                                if (typeSyntax.Kind() == SyntaxKind.QualifiedName)
+                                {
+                                    typeSyntax = ((QualifiedNameSyntax)typeSyntax).Right;
+                                }
+
+                                insideLoopExpression = (ExpressionSyntax listIdentifier, ExpressionSyntax expression) =>
+                                            SyntaxFactory.InvocationExpression(
+                                                SyntaxFactory.MemberAccessExpression(
+                                                    SyntaxKind.SimpleMemberAccessExpression,
+                                                    listIdentifier,
+                                                    SyntaxFactory.IdentifierName("Add")), // TODO get "Add" from nameof something
+                                                SyntaxFactory.ArgumentList(SyntaxFactory.SingletonSeparatedList(SyntaxFactory.Argument(expression)))
+                                                );
+                                initializer = SyntaxFactory.ObjectCreationExpression(
+                                    SyntaxFactory.GenericName(
+                                        SyntaxFactory.Identifier("List"),
```

Comment:
this will not work unless "using System.Collections.Generic" is in scope.  The better way to do this is to use SyntaxGenerator, or, get the type symbol you want, and call the extension .GenerateTypeSyntax on it.  It will do the right thing wrt to simplification.

Reply:
Thank you! I definitely should use the full qualifier.

I tried SyntaxGenerator. It actually returns and accepts SyntaxNode not specific type, e.g. ExpressionSyntax. So, it seems to be weakly typed.

So, I think that using SyntaxFactory will be consistent. Then, we still apply Simplifier to clean up namespaces if usings are already applied.

What do you think?

---
In reply to: [175267578](https://github.com/dotnet/roslyn/pull/25362#discussion_r175267578) [](ancestors = 175267578)

Reply:
> I tried SyntaxGenerator. It actually returns and accepts SyntaxNode not specific type, 

That's a pro.  Given taht it could let you use the same code for VB and C# potentially.

> So, I think that using SyntaxFactory will be consistent. Then, we still apply Simplifier to clean up namespaces if usings are already applied.

Unfortunately, that's not how it works.  The simplifier needs to know the original type symbols that created these types.  That's what things like generator.TypeExpression do for you.

> e.g. ExpressionSyntax. So, it seems to be weakly typed.

Note: in general you don't need to be strongly typed for this sort of feature.  If you do need a stsrong type in some cases, you can just cast as necessary.

Reply:
[https://github.com/dotnet/roslyn/issues/25639](https://github.com/dotnet/roslyn/issues/25639)

---
In reply to: [175617788](https://github.com/dotnet/roslyn/pull/25362#discussion_r175617788) [](ancestors = 175617788)